pkg/sql/logictest/tests/fakedist-vec-off/fakedist-vec-off_test: TestLogic_enums failed

pkg/sql/logictest/tests/fakedist-vec-off/fakedist-vec-off_test.TestLogic_enums [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5924055?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5924055?buildTab=artifacts#/) on master @ [5fe052748de0406d7cb2c8a1e9230143f54f7534](https://github.com/cockroachdb/cockroach/commits/5fe052748de0406d7cb2c8a1e9230143f54f7534):


```
=== RUN   TestLogic_enums/regression_63138
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_enums.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-18214

@ajwerner do you have an idea for why this diff
```
diff --git a/pkg/sql/catalog/typedesc/type_desc.go b/pkg/sql/catalog/typedesc/type_desc.go
index 02e77e5a40..08820572d9 100644
--- a/pkg/sql/catalog/typedesc/type_desc.go
+++ b/pkg/sql/catalog/typedesc/type_desc.go
@@ -796,6 +796,9 @@ func EnsureTypeIsHydrated(
                                return err
                        }
                }
+       } else if t.Family() == types.ArrayFamily {
+               // Hydrate the element type.
+               return EnsureTypeIsHydrated(ctx, t.ArrayContents(), res)
        }
        if !t.UserDefined() || t.IsHydrated() {
                return nil
@@ -897,11 +900,7 @@ func (desc *immutable) HydrateTypeInfoWithName(
        case descpb.TypeDescriptor_ALIAS:
                if typ.UserDefined() {
                        switch typ.Family() {
-                       case types.ArrayFamily:
-                               // Hydrate the element type.
-                               elemType := typ.ArrayContents()
-                               return EnsureTypeIsHydrated(ctx, elemType, res)
-                       case types.TupleFamily:
+                       case types.ArrayFamily, types.TupleFamily:
                                return EnsureTypeIsHydrated(ctx, typ, res)
                        default:
                                return errors.AssertionFailedf("unhandled alias type family %s", typ.Family())
```
would actually break things?
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4d265c5]

goroutine 4573 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc00024fdd0, {0x8cee300, 0xc0014a9890})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:248 +0xa5
panic({0x7ad82c0, 0xb33daa0})
	GOROOT/src/runtime/panic.go:838 +0x207
github.com/cockroachdb/cockroach/pkg/sql/types.(*T).PGName(0xc003df0240)
	github.com/cockroachdb/cockroach/pkg/sql/types/pkg/sql/types/types.go:1535 +0xc5
github.com/cockroachdb/cockroach/pkg/sql.glob..func109.1({0x8d546e0?, 0xc001779050?}, {0xc002b26588, 0x6}, {0x8d75c48, 0xc003def900})
	github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:530 +0xcb3
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDesc.func1({0x8d546e0?, 0xc001779050?}, {0xc002b26588?, 0x6800000069?}, {0x8d75c48?, 0xc003def900?}, 0xc001779050?)
	github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:2292 +0x37
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDescWithTableLookupInternalFromDescriptors({0x8cee300, 0xc0014a9890}, 0xc00164dc00, {0x8d546e0, 0xc0015d7320}, 0x0, 0x64?, {{{0xc003cb2080}, {0xc003cb2040}, 0xc003952db0}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:2549 +0x518
```

--- a/pkg/sql/catalog/typedesc/type_desc.go
+++ b/pkg/sql/catalog/typedesc/type_desc.go
@@ -877,41 +877,52 @@ func (desc *immutable) HydrateTypeInfoWithName(
 	if typ.IsHydrated() {
 		return nil
 	}
-	typ.TypeMeta.Name = &types.UserDefinedTypeName{
-		Catalog:        name.Catalog(),
-		ExplicitSchema: name.ExplicitSchema,
-		Schema:         name.Schema(),
-		Name:           name.Object(),
-	}
-	typ.TypeMeta.Version = uint32(desc.Version)
+	var enumData *types.EnumMetadata
 	switch desc.Kind {
 	case descpb.TypeDescriptor_ENUM, descpb.TypeDescriptor_MULTIREGION_ENUM:
 		if typ.Family() != types.EnumFamily {
 			return errors.New("cannot hydrate a non-enum type with an enum type descriptor")
 		}
-		typ.TypeMeta.EnumData = &types.EnumMetadata{
+		enumData = &types.EnumMetadata{
 			LogicalRepresentations:  desc.logicalReps,
 			PhysicalRepresentations: desc.physicalReps,
 			IsMemberReadOnly:        desc.readOnlyMembers,
 		}
-		return nil
 	case descpb.TypeDescriptor_ALIAS:
 		if typ.UserDefined() {
 			switch typ.Family() {
 			case types.ArrayFamily:
 				// Hydrate the element type.
 				elemType := typ.ArrayContents()
-				return EnsureTypeIsHydrated(ctx, elemType, res)
+				if err := EnsureTypeIsHydrated(ctx, elemType, res); err != nil {
+					return err
+				}
 			case types.TupleFamily:
-				return EnsureTypeIsHydrated(ctx, typ, res)
+				if err := EnsureTypeIsHydrated(ctx, typ, res); err != nil {
+					return err
+				}
 			default:
 				return errors.AssertionFailedf("unhandled alias type family %s", typ.Family())
 			}
 		}
-		return nil
 	default:
 		return errors.AssertionFailedf("unknown type descriptor kind %s", desc.Kind)
 	}
+
+	// Only hydrate the type if we did not fail to perform any of the above
+	// steps. If we were to populate these before something that might fail,
+	// the type may end up partially hydrated.
+	typ.TypeMeta = types.UserDefinedTypeMetadata{
+		Name: &types.UserDefinedTypeName{
+			Catalog:        name.Catalog(),
+			ExplicitSchema: name.ExplicitSchema,
+			Schema:         name.Schema(),
+			Name:           name.Object(),
+		},
+		Version:  uint32(desc.Version),
+		EnumData: enumData,
+	}
+	return nil
 }
 
 // NumEnumMembers implements the TypeDescriptor interface.

